Rick's PR of Ryans routing PR (#16456)
dotnet/AspNetCore.Docs: aspnetcore/mvc/controllers/routing.md

author: rick-anderson
description: Learn how ASP.NET Core MVC uses Routing Middleware to match URLs of incoming requests and map them to actions.
ms.author: riande

> ms.date: 1/25/2020

Reviewer: Something has gone awry with the sections, or delimiters. I'm seeing some sections duplicated. 

Also I expected to see the `IActionConstraint` section gone. We don't want to encourage people to use it.

![image](https://user-images.githubusercontent.com/1430011/74097527-b429c000-4ac1-11ea-8500-16f9775c5ba0.png)

Actions are either [conventionally-routed](#cr) or [attribute-routed](#ar). Placing a route on the controller or [action](#action) makes it attribute-routed. See [Mixed routing](#routing-mixed-ref-label) for more information.

This document:

* Explains the interactions between MVC and routing:
  * How typical MVC apps make use of routing features.
  * Covers both:
    * [Conventionally routing](#cr) typically used with controllers and views.
    * *Attribute routing* used with REST APIs. If you're primarily interested in routing for REST APIs, jump to the [Attribute routing for REST APIs](#ar) section.
  * See [Routing](xref:fundamentals/routing) for advanced routing details.
* Refers to the default routing system added in ASP.NET Core 3.0, called endpoint routing. It's possible to use controllers with the previous version of routing for compatibility purposes. See the [2.2-3.0 migration guide](xref:migration/22-to-30) for instructions. Refer to the [2.2 version of this document](https://docs.microsoft.com/aspnet/core/mvc/controllers/routing?view=aspnetcore-2.2) for reference material on the legacy routing system.

> ## Set up conventional endpoint routing middleware

Reviewer: I'd suggest changing this back to `Setting up Routing Middleware`

`conventional endpoint routing middleware` is not a term I want to use

Reviewer: Alternative: `Setting up a conventional route`

}
```

Using the preceding controller definition and route template, the `HomeController.Index` action is run for the following URL paths:

* `/Home/Index/17`
* `/Home/Index`
* `/Home`
* `/`

The URL path `/` uses the route template default `Home` controllers and `Index` action. The URL path `/Home` uses the route template default `Index` action.

The convenience method <xref:Microsoft.AspNetCore.Builder.ControllerEndpointRouteBuilderExtensions.MapDefaultControllerRoute*>:

```csharp
endpoints.MapDefaultControllerRoute();
```

Replaces:

```csharp
endpoints.MapControllerRoute("default", "{controller=Home}/{action=Index}/{id?}");
```

Routing is configured using the <xref:Microsoft.AspNetCore.Builder.EndpointRoutingApplicationBuilderExtensions.UseRouting*> and <xref:Microsoft.AspNetCore.Builder.EndpointRoutingApplicationBuilderExtensions.UseEndpoints*> middleware. To use controllers:

* Call <xref:Microsoft.AspNetCore.Builder.ControllerEndpointRouteBuilderExtensions.MapControllers*> inside `UseEndpoints` to map [attribute routed](#ar) controllers.

> * Call <xref:Microsoft.AspNetCore.Builder.ControllerEndpointRouteBuilderExtensions.MapControllerRoute*> or <xref:Microsoft.AspNetCore.Builder.ControllerEndpointRouteBuilderExtensions.MapAreaControllerRoute*>, to map [conventionally routed](#cr) controllers.

Reviewer: We should add a callout here to the effect of:

MVC's routing integration has two different programming experiences, conventional routing and attribute routing. There is one common routing implementation that is used regardless of the way in which it's programmed. Conventional routing concepts and attribute routing concepts are used to program the underlying routing system.

Reviewer: Nevermind. This is already here later in the doc.

<!-- review: Why would you ever use a dedicated conventional route like this? -->
<a name="dcr"></a>

The `blog` route in the preceding code is a **dedicated conventional route**. It's called a dedicated conventional route because:

* It uses the [conventional routing](#cr) system.
* It's dedicated to a specific [action](#action).

Because `controller` and `action` don't appear in the route template `"blog/{*article}"` as parameters:

* They can only have the default values `{ controller = "Blog", action = "Article" }`.
* This route will always map to the action `BlogController.Article`.

`/Blog`, `/Blog/Article`, and `/Blog/{any-string}` are the only URL paths that match the blog route.

In the preceding example, the `blog` route has a higher priority for matches than the `default` route.

<a name="greedy"></a>

[Dedicated conventional routes](#dcr) often use catch all route parameters like `{*article}` to capture the remaining portion of the URL path. This can make a route too **greedy**, meaning that it matches URLs that you intended to be matched by other routes. Put the greedy routes later in the route table to solve this.

<!-- Review: 

> Should I change:  In ASP.NET Core 3.0 and later, the routing system doesn't:

Reviewer: No.

The URL path `/` uses the route template default `Home` controllers and `Index` action. The URL path `/Home` uses the route template default `Index` action.

The convenience method <xref:Microsoft.AspNetCore.Builder.ControllerEndpointRouteBuilderExtensions.MapDefaultControllerRoute*>:

```csharp
endpoints.MapDefaultControllerRoute();
```

Replaces:

```csharp
endpoints.MapControllerRoute("default", "{controller=Home}/{action=Index}/{id?}");
```

Routing is configured using the <xref:Microsoft.AspNetCore.Builder.EndpointRoutingApplicationBuilderExtensions.UseRouting*> and <xref:Microsoft.AspNetCore.Builder.EndpointRoutingApplicationBuilderExtensions.UseEndpoints*> middleware. To use controllers:

* Call <xref:Microsoft.AspNetCore.Builder.ControllerEndpointRouteBuilderExtensions.MapControllers*> inside `UseEndpoints` to map [attribute routed](#ar) controllers.
* Call <xref:Microsoft.AspNetCore.Builder.ControllerEndpointRouteBuilderExtensions.MapControllerRoute*> or <xref:Microsoft.AspNetCore.Builder.ControllerEndpointRouteBuilderExtensions.MapAreaControllerRoute*>, to map [conventionally routed](#cr) controllers.

<a name="routing-conventional-ref-label"></a>
<a name="cr"></a>

> ## Conventional endpoint routing

Reviewer: This should be reverted to `Conventional routing` (here and everywhere else it appears. `Conventional endpoint routing` is not a term I want to use.

Conventional endpoint routing is used with controllers and views. The `default` route:

[!code-csharp[](routing/samples/3.x/main/StartupDefaultMVC.cs?name=snippet2)]

is an example of a *conventional routing*. It's called *conventional routing* because it establishes a *convention* for URL paths:

* The first path segment, `{controller=Home}`, maps to the controller name.
* The second segment, `{action=Index}`, maps to the [action](#action) name.
* The third segment, `{id?}` is used for an optional `id`. The `?` in `{id?}` makes it optional. `id` is used to map to a model entity.

Using this `default` route, the URL path:

* `/Products/List` maps to the `ProductsController.List` action.
* `/Blog/Article/17` maps to `BlogController.Article` and typically model binds the `id` parameter to 17.

This mapping:

* Is based on the controller and [action](#action) names **only**.
* Isn't based on namespaces, source file locations, or method parameters.

> * Is different than typical [attribute routing](#ar) which doesn't include the action name in the route.

Reviewer: I suggest removing this point. There are lots of cases where users choose to include the action name in the route using attribute routing.

Using conventional routing with the default route allows creating the app without having to come up with a new URL pattern for each action. For an app with [CRUD](https://wikipedia.org/wiki/Create,_read,_update_and_delete) style actions, having consistency for the URLs across controllers:

* Helps simplify the code.
* Makes the UI more predictable.

Most apps should choose a basic and descriptive routing scheme so that URLs are readable and meaningful. The default conventional route `{controller=Home}/{action=Index}/{id?}`:

* Supports a basic and descriptive routing scheme.
* Is a useful starting point for UI-based apps.
* Is the only route template needed for many web UI apps. For larger web UI apps, another route using [Areas](#areas) if frequently all that's needed.

<xref:Microsoft.AspNetCore.Builder.ControllerEndpointRouteBuilderExtensions.MapControllerRoute*> and <xref:Microsoft.AspNetCore.Builder.MvcAreaRouteBuilderExtensions.MapAreaRoute*> :

* Automatically assign an **order** value to their endpoints based on the order they are invoked.

> * Don't expose the [RouteEndpoint.Order](xref:Microsoft.AspNetCore.Routing.RouteEndpoint.Order*). `RouteEndpoint.Order` is only exposed for [attribute routing](#ar). <!-- review required for preceding two bullets. Are they OK? -->

Reviewer: I'm not sure this is the right way to explain this point. Maybe rather than describing attribute routing + order here, leave it for the attribute routing section?

I think the first bullet point is good.

Endpoint routing in ASP.NET Core 3.0 and later:

* Doesn't have a concept of routes.

> * Doesn't provide ordering guarantees, all endpoints are processed at once.

Reviewer: ```suggestion
* Doesn't provide ordering guarantees for the execution of extensibility,  all endpoints are processed at once.
```

> [!WARNING]

> > The `id` is defined as optional by the preceding route template. Actions can execute without the optional ID provided as part of the URL. Generally, when`id` is omitted from the URL:

Reviewer: This warning needs to appear closer to the example, it's totally out of context here

>
> [Attribute routing](#ar) provides fine-grained control to make the ID required for some actions and not for others. By convention, the documentation includes optional parameters like `id` when they're likely to appear in correct usage.

Enable [Logging](xref:fundamentals/logging/index) to see how the built-in routing implementations, such as <xref:Microsoft.AspNetCore.Routing.Route>, match requests.

[Attribute routing](#ar) is explained later in this document.

<a name="mr"></a>

### Multiple conventional routes

Multiple [conventional routes](#cr) can be added inside `UseEndpoints` by adding more calls to <xref:Microsoft.AspNetCore.Builder.ControllerEndpointRouteBuilderExtensions.MapControllerRoute*> and <xref:Microsoft.AspNetCore.Builder.ControllerEndpointRouteBuilderExtensions.MapAreaControllerRoute*>. Doing so allows defining multiple conventions, or to adding conventional routes that are dedicated to a specific [action](#action), such as:

[!code-csharp[](routing/samples/3.x/main/Startup.cs?name=snippet_1)]

<a name="dcr"></a>

The `blog` route in the preceding code is a **dedicated conventional route**. It's called a dedicated conventional route because:

> * It uses the [conventional routing](#cr) system.

Reviewer: We should not describe conventional routing (nor attribute routing) as a **system**. There is one routing system, and those are the ways of configuring it. 

I'd suggest dropping the word `system`. 

(this is a mistake I should have caught in my rewrite, legacy routing kinda had two systems)

* It's dedicated to a specific [action](#action).

Because `controller` and `action` don't appear in the route template `"blog/{*article}"` as parameters:

* They can only have the default values `{ controller = "Blog", action = "Article" }`.
* This route always maps to the action `BlogController.Article`.

`/Blog`, `/Blog/Article`, and `/Blog/{any-string}` are the only URL paths that match the blog route.

The preceding example:

> * `blog` route has a higher priority for matches than the `default` route.

Reviewer: ```suggestion
* `blog` route has a higher priority for matches than the `default` route because it is added first.
```

Conventional routing only matches a combination of action and controller that are defined by the app. This is intended to simplify cases where conventional routes overlap.
Adding routes using <xref:Microsoft.AspNetCore.Builder.ControllerEndpointRouteBuilderExtensions.MapControllerRoute*>, <xref:Microsoft.AspNetCore.Builder.ControllerEndpointRouteBuilderExtensions.MapDefaultControllerRoute*>, and <xref:Microsoft.AspNetCore.Builder.ControllerEndpointRouteBuilderExtensions.MapAreaControllerRoute*> automatically assign an order value to their endpoints based on the order they are invoked. Matches from a route that appears earlier have a higher priority. Conventional routing is order-dependent. In general, routes with areas should be placed earlier as they're more specific than routes without an area. [Dedicated conventional routes](#dcr) with catch all route parameters like `{*article}` can make a route too [greedy](xref:fundamentals/routing#greedy), meaning that it matches URLs that you intended to be matched by other routes. Put the greedy routes later in the route table to prevent greedy matches.

<a name="cctar"></a>

### ControllerContext.ToActionResult

> Many of the samples in this document use the following `ControllerContextExtensions`class to display <xref:Microsoft.AspNetCore.Mvc.Abstractions.ActionDescriptor> information:

Reviewer: I think it's weird to show this code here because it deals with a bunch of APIs that are not described like `ControllerContext` and `ActionDescriptor`. Would it be appropriate to just link to the code?

Reviewer: I would also *****REALLYY***** like to see this take a different shape than an extension method on a framework type called `ToActionResult`. This makes it really easy to forget that this isn't part of the product, I think it's artificially look built-in rather than make it clear that it's something we're using to explain the product.

In the following example:

* The preceding `Configure` method is used.
* `HomeController` matches a set of URLs similar to what the default conventional route `{controller=Home}/{action=Index}/{id?}` matches.

[!code-csharp[](routing/samples/3.x/main/Controllers/HomeController.cs?name=snippet2)]

The `HomeController.Index` action is run for any of the URL paths `/`, `/Home`, `/Home/Index`, or `/Home/Index/3`.

This example highlights a key programming difference between attribute routing and [conventional routing](#cr). Attribute routing requires more input to specify a route. The conventional default route handles routes more succinctly. However, attribute routing allows and requires precise control of which route templates apply to each [action](#action).

With attribute routing, the controller name and action names play **no** role in which action is matched. The following example matches the same URLs as the previous example:

[!code-csharp[](routing/samples/3.x/main/Controllers/MyDemoController.cs?name=snippet)]

> The following code uses route parameters for `action` and `controller`:

Reviewer: ```suggestion
The following code uses token replacement for `action` and `controller`:
```

They aren't parameters 😆

* `controller`
* `handler`
* `page`

Using `page` as a route parameter with attribute routing is a common error. Doing that results in inconsistent and confusing behavior with URL generation.

[!code-csharp[](routing/samples/3.x/main/Controllers/MyDemo2Controller.cs?name=snippet)]

The special parameter names are used by the URL generation system to determine if a URL generation operation refers to a Razor Page or to a Controller.

<a name="verb"></a>

## HTTP verb templates

ASP.NET Core has the following HTTP verb templates:

* [[HttpGet]](xref:Microsoft.AspNetCore.Mvc.HttpGetAttribute)
* [[HttpPost]](xref:Microsoft.AspNetCore.Mvc.HttpPostAttribute)
* [[HttpPut]](xref:Microsoft.AspNetCore.Mvc.HttpPutAttribute)
* [[HttpDelete]](xref:Microsoft.AspNetCore.Mvc.HttpDeleteAttribute)
* [[HttpHead]](xref:Microsoft.AspNetCore.Mvc.HttpHeadAttribute)
* [[HttpPatch]](xref:Microsoft.AspNetCore.Mvc.HttpPatchAttribute)

<a name="rt"></a>

### Route templates

ASP.NET Core has the following route templates:

* All the [HTTP verb templates](#verb) are route templates.
* [[Route]](xref:Microsoft.AspNetCore.Mvc.RouteAttribute)

<a name="arx"></a>

### Attribute routing with Http verb attributes

> Unlike routes created for controllers with views that generally include `[action]` in the route template, API controllers typically don't include `[action]` in the route template. Because `[action]` isn't in the route template, the [action](#action) name is not in the route. That is, the method name isn't used in the matching route.

Reviewer: I'm not super sure what this is trying to explain and why it's so important. We literally have a feature that does this, and we just showed its use. 

I don't think we need to discourage people from doing this, I can think of cases where I would use it even in a RESTful API.

[!code-csharp[](routing/samples/3.x/main/Controllers/ProductsApiController.cs?name=snippet)]

In the preceding example:

* The URL path `/products` can match `ProductsApi.ListProducts`
* The URL path `/products/5` can match `ProductsApi.GetProduct(int)`.

Both of these actions only match HTTP `GET` because they're marked with the `[HttpGet]` attribute.

Route templates applied to an action that begin with `/` or `~/` don't get combined with route templates applied to the controller. The following example matches a set of URL paths similar to the default route.

[!code-csharp[](routing/samples/3.x/main/Controllers/HomeController.cs?name=snippet)]

The following table explains the `[Route]` attributes in the preceding code:

| Attribute               | Combines with `[Route("Home")]` | Defines route template |
| ----------------- | ------------ | --------- |
| `[Route("")]` | Yes | `"Home"` |
| `[Route("Index")]` | Yes | `"Home/Index"` |
| `[Route("/")]` | **No** | `""` |
| `[Route("About")]` | Yes | `"Home/About"` |

<a name="routing-ordering-ref-label"></a>
<a name="oar"></a>

### Attribute route order

> Attribute routing builds a tree and matches all routes simultaneously:

Reviewer: ```suggestion 
Routing builds a tree and matches all endpoints simultaneously:
```

```text
AmbiguousMatchException: The request matched multiple endpoints. Matches:

 WebMvcRouting.Controllers.HomeController.Index
 WebMvcRouting.Controllers.MyDemoController.MyIndex
```

Adding `Order` to one of the route attributes resolves the ambiguity:

[!code-csharp[](routing/samples/3.x/main/Controllers/MyDemo3Controller.cs?name=snippet3& highlight=2)]

With the preceding code, `/home` runs the `HomeController.Index` endpoint. To get to the `MyDemoController.MyIndex`, request `/home/MyIndex`. **Note**:

* The preceding code is an example or poor routing design. It was used to illustrate the `Order` property.
* The `Order` property only resolves the ambiguity, that template cannot be matched. It would be better to remove the `[Route("Home")]` template.

See [Razor Pages route and app conventions: Route order](xref:razor-pages/razor-pages-conventions#route-order) for information on route order with Razor Pages.

In some cases, an HTTP 500 error is returned with ambiguous routes. Use [logging](xref:fundamentals/logging/index) to see which endpoints caused the the `AmbiguousMatchException`.

<a name="routing-token-replacement-templates-ref-label"></a>

> ## Token replacement in route templates [controller], [action], [area]

Reviewer: It's possible that this heading is at the wrong level. Things are nested under this node in the TOC that are unrelated.

Reviewer: I'm not sure, I guess I expected to see everything in TOC, but my expectations were wrong 😆

With attribute routing, the controller name and action names play **no** role in which action is matched. The following example matches the same URLs as the previous example:

[!code-csharp[](routing/samples/3.x/main/Controllers/MyDemoController.cs?name=snippet)]

The following code uses route parameters for `action` and `controller`:

[!code-csharp[](routing/samples/3.x/main/Controllers/HomeController.cs?name=snippet22)]

The following code applies `[Route("[controller]/[action]")]` to the controller:

[!code-csharp[](routing/samples/3.x/main/Controllers/HomeController.cs?name=snippet24)]

In the preceding code, the `Index` method templates must prepend `/` or `~/` to the route templates. Route templates applied to an action that begin with `/` or `~/` don't get combined with route templates applied to the controller.

## Reserved routing names

The following keywords are reserved route parameter names when using Controllers or Razor Pages:

* `action`
* `area`
* `controller`
* `handler`
* `page`

Using `page` as a route parameter with attribute routing is a common error. Doing that results in inconsistent and confusing behavior with URL generation.

> <!-- review notes: Your template was missing terminating `"` [Route("/articles/{page})] that's why we use snippets :) -->

Reviewer: Cool beans

* [[HttpGet]](xref:Microsoft.AspNetCore.Mvc.HttpGetAttribute)
* [[HttpPost]](xref:Microsoft.AspNetCore.Mvc.HttpPostAttribute)
* [[HttpPut]](xref:Microsoft.AspNetCore.Mvc.HttpPutAttribute)
* [[HttpDelete]](xref:Microsoft.AspNetCore.Mvc.HttpDeleteAttribute)
* [[HttpHead]](xref:Microsoft.AspNetCore.Mvc.HttpHeadAttribute)
* [[HttpPatch]](xref:Microsoft.AspNetCore.Mvc.HttpPatchAttribute)

<a name="rt"></a>

### Route templates

ASP.NET Core has the following route templates:

* All the [HTTP verb templates](#verb) are route templates.
* [[Route]](xref:Microsoft.AspNetCore.Mvc.RouteAttribute)

<a name="arx"></a>

## Attribute routing with Http verb attributes

<!-- I had to add this sample code/explanation, even though you can get it elsewhere in the doc. But the doc doesn't put it all together in one place. I need an H2 to link to this information because we get so many issues on https://docs.microsoft.com/en-us/aspnet/core/tutorials/first-web-api where they don't understand constraints and how the route is determined. I may move this section to an include and add it in the Web API tutorial - Or I could move it there and not include it here.

> For example, https://github.com/aspnet/AspNetCore.Docs/issues/16901

Reviewer: I'm not totally sure what you're trying to say - what's the point of confusion? The concatenation of templates? What would help alleviate it? Better conceptual explanation? Better examples?

I'm fine with what you added here 👍

  * Doesn't match this action.
  * Returns a [404 Not Found](https://developer.mozilla.org/docs/Web/HTTP/Status/404) error.
    [!code-csharp[](routing/samples/3.x/main/Controllers/Test2Controller.cs?name=snippet3)]
* The `GetInt2Product` action contains `{id}` in the template, but doesn't constrain `id` to values that can be converted to an integer. A GET request to `/api/test2/int2/abc`:
  * Matches this route.
  * Model binding fails to convert `abc` to an integer. The `id` parameter of the method is integer.
  * Returns a [400 Bad Request](https://developer.mozilla.org/docs/Web/HTTP/Status/400) because model binding failed to convert`abc` to an integer.
      [!code-csharp[](routing/samples/3.x/main/Controllers/Test2Controller.cs?name=snippet4)]

Attribute routing can use <xref:Microsoft.AspNetCore.Mvc.Routing.HttpMethodAttribute> attributes such as <xref:Microsoft.AspNetCore.Mvc.HttpPostAttribute>, <xref:Microsoft.AspNetCore.Mvc.HttpPutAttribute>, and <xref:Microsoft.AspNetCore.Mvc.HttpDeleteAttribute>. All of the [HTTP verb](#verb) attributes accept a route template. The following example shows two actions that match the same route template:

> <!-- review: My snippets use :ControllerBase, you used : Controller

Reviewer: I tend to use `Controller` for views and `ControllerBase` for non-views.

`ActionResult` > `IActionResult`

[!code-csharp[](routing/samples/3.x/main/Controllers/ProductsApiController.cs?name=snippet)]

In the preceding example:

* The URL path `/products` can match `ProductsApi.ListProducts`
* The URL path `/products/5` can match `ProductsApi.GetProduct(int)`.

Both of these actions only match HTTP `GET` because they're marked with the `[HttpGet]` attribute.

Route templates applied to an action that begin with `/` or `~/` don't get combined with route templates applied to the controller. The following example matches a set of URL paths similar to the default route.

[!code-csharp[](routing/samples/3.x/main/Controllers/HomeController.cs?name=snippet)]

The following table explains the `[Route]` attributes in the preceding code:

| Attribute               | Combines with `[Route("Home")]` | Defines route template |
| ----------------- | ------------ | --------- |
| `[Route("")]` | Yes | `"Home"` |
| `[Route("Index")]` | Yes | `"Home/Index"` |
| `[Route("/")]` | **No** | `""` |
| `[Route("About")]` | Yes | `"Home/About"` |

<a name="routing-ordering-ref-label"></a>
<a name="oar"></a>

### Attribute route order

> <!-- In the  original, you had "In contrast to conventional routes which execute in a defined order"

Reviewer: If there are still references to this we should remove them. Conventional routes define a "matching priority", not an "execution order"

> Is that no longer true? If that's still true need to keep that and add that to the #cr section.

Reviewer: > Is that no longer true?

It is no longer true on SO many levels:

- There is no such thing as a route
- There is nothing to execute because there is no such thing as a route
- We do not provide any guarantees about execution order, only matching priority :)

`[Route("[controller]/[action]", Name="[controller]_[action]")]`generates a unique route name for each action:

[!code-csharp[](routing/samples/3.x/main/Controllers/ProductsController.cs?name=snippet5)]

In the preceding code:

* The `List` action returns the route name `products_list`.
* The `Edit` action returns a `null` route name. The `Edit` route name is null because the route is `api/Products/Edit/id`, which doesn't match `api/[controller]/[action]`.

To match the literal token replacement delimiter `[` or  `]`, escape it by repeating the character (`[[` or `]]`).

<a name="routing-token-replacement-transformers-ref-label"></a>

### Use a parameter transformer to customize token replacement

Token replacement can be customized using a parameter transformer. A parameter transformer implements <xref:Microsoft.AspNetCore.Routing.IOutboundParameterTransformer> and transforms the value of parameters. For example, a custom `SlugifyParameterTransformer` parameter transformer changes the `SubscriptionManagement` route value to `subscription-management`:

> <!-- review: Changed  ToLower to ToLowerInvariant per https://github.com/dotnet/aspnetcore/issues/18774 -->

Reviewer: 👍

* Customizes the attribute route token values as they are replaced.

[!code-csharp[](routing/samples/3.x/main/Controllers/SubscriptionManagementController.cs?name=snippet)]

The preceding `ListAll` method matches `/subscription-management/list-all`.

The `RouteTokenTransformerConvention` is registered as an option in `ConfigureServices`.

[!code-csharp[](routing/samples/3.x/main/StartupSlugifyParamTransformer.cs?name=snippet)]

See [MDN web docs on Slug](https://developer.mozilla.org/docs/Glossary/Slug) for the definition of Slug.

<a name="routing-multiple-routes-ref-label"></a>

### Multiple attribute routes

Attribute routing supports defining multiple routes that reach the same action. The most common usage of this is to mimic the behavior of the default conventional route as shown in the following example:

[!code-csharp[](routing/samples/3.x/main/Controllers/ProductsController.cs?name=snippet6x)]

Putting multiple route attributes on the controller means that each one will combine with each of the route attributes on the action methods:

[!code-csharp[](routing/samples/3.x/main/Controllers/ProductsController.cs?name=snippet6)]

> When multiple route attributes that implement <xref:Microsoft.AspNetCore.Mvc.ActionConstraints.IActionConstraint> are placed on an action, each action constraint combines with the route template from the attribute that defined it. <!-- What is it?? Sentence is really too long to MT -->  All the [HTTP verb](#verb) route constraints implement `IActionConstraint`. <!-- Review/edit and remove comment marks. For example, in the preceding code, the post constraint `[HttpPost("Buy")]` combines with -->

Reviewer: Dunno what MT means. Seems important. Manual Transmission? Mountain? Montana?

In the preceding code:

* The base `namespace` is `My.Application`.
* The full name of the preceding controller is `My.Application.Admin.Controllers.UsersController`.
* The `NamespaceRoutingConvention` sets the controllers template to `Admin/Controllers/Users/[action]/{id?`.

The `NamespaceRoutingConvention` can also be applied as an attribute on a controller:

[!code-csharp[](routing/samples/3.x/nsrc/Controllers/TestController.cs?name=snippet&highlight=1)]

<a name="routing-mixed-ref-label"></a>

## Mixed routing: Attribute routing vs conventional routing

ASP.NET Core apps can mix the use of conventional routing and attribute routing. It's typical to use conventional routes for controllers serving HTML pages for browsers, and attribute routing for controllers serving REST APIs.

Actions are either conventionally routed or attribute routed. Placing a route on the controller or the action makes it attribute routed. Actions that define attribute routes cannot be reached through the conventional routes and vice-versa. **Any** route attribute on the controller makes **all** actions in the controller attribute routed.

> <!-- review: Can we delete based on the historical use cases for each system. 

Reviewer: Who's Dan Roth?

Dan.Roth does'nt really care about history. It's what we have today.
-->

> Attribute routing and conventional routing use the same routing engine. They differ in details of how URL generation works based on the historical use cases for each system. This different is expressed by how MVC interfaces with and provides data to the routing system.

Reviewer: ```suggestion
Attribute routing and conventional routing use the same routing engine.
```

Reviewer: The rest of these details aren't really true.

TagHelpers generate URLs through the `form` TagHelper and the `<a>` TagHelper. Both of these use `IUrlHelper` for their implementation. See [Tag Helpers in forms](xref:mvc/views/working-with-forms) for more information.

Inside views, the `IUrlHelper` is available through the `Url` property for any ad-hoc URL generation not covered by the above.

<a name="routing-gen-urls-action-ref-label"></a>

### URL generation in Action Results

The preceding examples showed using `IUrlHelper` in a controller. The most common usage in a controller is to generate a URL as part of an action result.

The <xref:Microsoft.AspNetCore.Mvc.ControllerBase> and <xref:Microsoft.AspNetCore.Mvc.Controller> base classes provide convenience methods for action results that reference another action. One typical usage is to redirect after accepting user input:

[!code-csharp[](routing/samples/3.x/main/Controllers/CustomerController.cs?name=snippet)]

> <!-- review: What are the action results factory methods? -->

Reviewer: Good point. These are things like `RedirectToAction` and `CreatedAtAction`

[!code-csharp[](routing/samples/3.x/AreasRouting/Startup.cs?name=snippet1)]

In the preceding code, <xref:Microsoft.AspNetCore.Builder.ControllerEndpointRouteBuilderExtensions.MapAreaControllerRoute*> is called to create the `"blog_route"`. The second parameter, `"Blog"`, is the area name.

When matching a URL path like `/Manage/Users/AddUser`, the `"blog_route"` route generates the route values `{ area = Blog, controller = Users, action = AddUser }`. The `area` route value is produced by a default value for `area`. The route created by `MapAreaControllerRoute` is equivalent to the following:

[!code-csharp[](routing/samples/3.x/AreasRouting/Startup2.cs?name=snippet2)]

`MapAreaControllerRoute` creates a route using both a default value and constraint for `area` using the provided area name, in this case `Blog`. The default value ensures that the route always produces `{ area = Blog, ... }`, the constraint requires the value `{ area = Blog, ... }` for URL generation.

Conventional routing is order-dependent. In general, routes with areas should be placed earlier as they're more specific than routes without an area.

> Usin

Reviewer: Not really. It's an attempt to simplify a really complex concept. let's just take it out.

> Usin

Reviewer: Can you be more specific?

### Ordering attribute routes

> In contrast to conventional routes, which execute in a defined order, attribute routing builds a tree and matches all routes simultaneously. This behaves as-if the route entries were placed in an ideal ordering; the most specific routes have a chance to execute before the more general routes.

Reviewer: Leaving this alone because my understanding is that this is the pre-3.0 docs.
